Automattic/wp-calypso
# Add Restore DNS options for our email services

#### Changes proposed in this Pull Request

* This PR explores adding a general-purpose way to restore the correct DNS records for the email services we offer: Email Forwarding, Google Workspace/G Suite, and Professional Email
* As per discussion prompted by an early exploration, we're handling this via the restore/action menu on the DNS configuration page
* Also, as per [this comment](https://github.com/Automattic/wp-calypso/pull/60709#issuecomment-1028016982) and [this comment](https://github.com/Automattic/wp-calypso/pull/60709#issuecomment-1028007278), we're only showing the button if the user has that service and we definitively know that the user doesn't have the expected DNS records via the new flags added in D74409-code (we're also leveraging the new mail records-only template that was added in that change).
   - For now, the Titan menu option is disabled for subdomains, as we have some other bugs to address with how that template is handled for subdomains
* I also made a change to disable all of the restore buttons if the domain object is not currently known (which is most commonly the case while loading the page) -- this should help prevent unintentional DNS changes that aren't actually necessary.
* I am not sure how to create a good test case for Email Forwarding, as the API call to remove the email forwarding MX record also removes all email forwards

#### Testing instructions

* Run this branch locally or via [calypso.live](https://github.com/Automattic/wp-calypso/pull/60709#issuecomment-1026988290)
* For domains (or domain mappings) you own with various email services (i.e. Google Workspace/G Suite, Professional Email, or Email Forwarding), navigate to Upgrades -> Domains -> <your-domain.com>
* Expand the "DNS records" section, and click on the "Manage" button
* Click on the three-dot menu in the top right corner
* Verify that the Email Forwarding menu item is not visible
* Verify that the Professional Email menu item is not visible
* Verify that the Google Workspace menu item is not visible

Now, for domains that each have a subscription with Titan or Google, follow these steps:
* Navigate to Upgrades -> Domains -> <your-domain.com>
* Expand the "DNS records" section, and click on the "Manage" button
* Click on the three-dot menu in the top right corner
* Verify that all three email restore options are not visible
* Update the DNS to remove an `MX` record or remove a `TXT` record that starts with `v=spf1` and includes the email provider's service name
* Refresh the page to trigger a new API call to the domains end point
  - During one of your tests verify that none of the options in the top right menu, including the "Restore default A records" button, are visible while we're fetching data from the server
* Click on the three-dot menu in the top right corner
* Verify that the restore button for the current service is now visible, and click on it
* Verify that you see the confirmation dialog
* Verify that the record(s) you deleted are added back if you click on "Restore" and they are not if you click on "Cancel"
* Verify that clicking on either of the buttons results in the dialog being removed

For extra credit, verify that we disable the Professional Email option when dealing with a subdomain, even if it has a Professional Email subscription and invalid DNS records.

##### Screenshots

![image](https://user-images.githubusercontent.com/5689927/153593623-d67b5523-4255-4a0d-a0b6-5475a880db58.png)
<img width="1134" alt="Screenshot 2022-02-02 at 11 47 08" src="https://user-images.githubusercontent.com/3376401/152130409-1c84757c-94f9-49f8-a40b-a5078c3343bb.png">
